onnxruntime/core/providers/cuda/cuda_graph.cc

```diff
@@ -29,40 +73,70 @@ void CUDAGraph::CaptureBegin() {
   CUDA_CALL_THROW(cudaStreamBeginCapture(stream_, cudaStreamCaptureModeGlobal));
 }
 
-void CUDAGraph::CaptureEnd() {
+void CUDAGraphManager::CaptureEnd() {
+  if (!IsGraphCaptureAllowedOnRun()) {
+    LOGS_DEFAULT(INFO) << "Skipping graph capture for cuda_graph_annotation_id " << cuda_graph_annotation_id_;
+    return;
+  }
+
   CUDA_CALL_THROW(cudaStreamEndCapture(stream_, &graph_));
   if (graph_ == NULL) {
     ORT_THROW("CUDAGraph::CaptureEnd: graph_ is NULL");
   }
-
   has_graph_ = true;
-  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
-  has_graph_exec_ = true;
+
+  cudaGraphExec_t graph_exec = NULL;
+  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec, graph_, NULL, NULL, 0));
   CUDA_CALL_THROW(cudaGraphDestroy(graph_));
   has_graph_ = false;
+
+  cuda_graph_set_.Put(cuda_graph_annotation_id_, graph_exec);
+  has_graph_exec_ = true;
 }
 
-Status CUDAGraph::Replay() {
+Status CUDAGraphManager::Replay() {
+  if (!IsGraphCaptureAllowedOnRun()) {
+    LOGS_DEFAULT(INFO) << "Skipping graph replay for cuda_graph_annotation_id " << cuda_graph_annotation_id_;
+    return Status::OK();
+  }
   // Although this function is not thread safe, the lock is not needed here because
   // CUDA EP maintains a separate cuda graph per thread
-  LOGS_DEFAULT(INFO) << "Replaying CUDA graph on stream " << stream_;
-  CUDA_RETURN_IF_ERROR(cudaGraphLaunch(graph_exec_, stream_));
+  LOGS_DEFAULT(INFO) << "Replaying CUDA graph on stream " << stream_ << " with cuda_graph_annotation_id " << cuda_graph_annotation_id_;
+
+  cudaGraphExec_t graph_exec = NULL;
+  if (!cuda_graph_set_.Get(cuda_graph_annotation_id_, graph_exec)) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME,
+                           FAIL,
+                           "CUDAGraph::Replay: cuda_graph_set_ does not contain the cuda_graph_annotation_id");
+  }
+  CUDA_RETURN_IF_ERROR(cudaGraphLaunch(graph_exec, stream_));
+
   CUDA_RETURN_IF_ERROR(cudaStreamSynchronize(stream_));
   return Status::OK();
 }
 
-void CUDAGraph::Reset() {
+bool CUDAGraphManager::IsGraphCaptureAllowedOnRun() const {
+  return cuda_graph_annotation_id_ != kCudaGraphAnnotationSkip;
+}
+
+bool CUDAGraphManager::IsGraphCaptured(CudaGraphAnnotation_t cuda_graph_annotation_id) const {
+  if (!IsGraphCaptureAllowedOnRun()) {
+    return false;
+  }
+  return cuda_graph_set_.Contains(cuda_graph_annotation_id);
+}
+
+void CUDAGraphManager::Reset() {
   if (has_graph_) {
     CUDA_CALL_THROW(cudaGraphDestroy(graph_));
     has_graph_ = false;
   }
-  if (has_graph_exec_) {
-    CUDA_CALL_THROW(cudaGraphExecDestroy(graph_exec_));
-    has_graph_exec_ = false;
+  if (!cuda_graph_set_.IsEmpty()) {
```

Comment:
In my opinion, we don't need `IsEmpty()` - we can just handle all that in `Clear()` (do nothing in Clear is it already empty) - Do you foresee any future use for `IsEmpty()` ?